File: backend/shop/serializers.py
from rest_framework import serializers
from .models import Product, Sale


class ProductSerializer(serializers.ModelSerializer):
    image_url = serializers.SerializerMethodField()

    class Meta:
        model = Product
        fields = [
            'id', 'name', 'category', 'description',
            'price', 'stock', 'image', 'image_url',
            'created_at', 'updated_at',
        ]
        read_only_fields = ['id', 'created_at', 'updated_at']

    def get_image_url(self, obj):
        if obj.image:
             url = str(obj.image.url)
        return url.replace('http://', 'https://')  # Ensure secure URL
        return None


class SaleSerializer(serializers.ModelSerializer):
    class Meta:
        model = Sale
        fields = [
            'id', 'product', 'product_name', 'product_category',
            'unit_price', 'quantity', 'total', 'sold_at',
        ]
        read_only_fields = [
            'id', 'product_name', 'product_category',
            'unit_price', 'total', 'sold_at',
        ]


class SellSerializer(serializers.Serializer):
    quantity = serializers.IntegerField(min_value=1)

    def validate_quantity(self, value):
        product = self.context.get('product')
        if product and value > product.stock:
            raise serializers.ValidationError(
                f"Only {product.stock} units in stock."
            )
        return value


class RestockSerializer(serializers.Serializer):
    quantity = serializers.IntegerField(min_value=1)

File: lib/serializers.js
// Shape responses exactly like the old Django REST API so the
// frontend logic carries over unchanged.

export function serializeProduct(p) {
  return {
    id: p.id,
    name: p.name,
    category: p.category,
    description: p.description,
    price: p.price.toString(),
    stock: p.stock,
    image: p.image,
    image_url: p.image ? p.image.replace('http://', 'https://') : null,
    created_at: p.createdAt.toISOString(),
    updated_at: p.updatedAt.toISOString(),
  }
}

export function serializeSale(s) {
  return {
    id: s.id,
    product: s.productId,
    product_name: s.productName,
    product_category: s.productCategory,
    unit_price: s.unitPrice.toString(),
    quantity: s.quantity,
    total: s.total.toString(),
    sold_at: s.soldAt.toISOString(),
  }
}
